include/seastar/core/iostream-impl.hh

```diff
@@ -198,8 +202,21 @@ input_stream<CharType>::read_exactly(size_t n) noexcept {
     } else {
       try {
         // buffer too small: start copy/read loop
-        tmp_buf b(n);
-        return read_exactly_part(n, std::move(b), 0);
+        tmp_buf buf(n);
+        size_t count {0U};
+        size_t avail = available();
+        if (avail) {
```

Comment:
is this branch actually good to have?
(we can also simply compute `count`, and call copy_n and trim_front, even when `count == 0`.)

Reply:
I think I gave you bad advice re the split - all the no-allocation cases are covered in read_exactly(). If we ever drop into read_exactly_part(), that means we'll allocate for sure (e.g. we won't hit `if (completed == n)` in the first iteration. So this patch is unnecessary.

Please verify and drop if I'm right in this second iteration.

Reply:
I've dropped the unnecessary commit. Added a new commit where I do some more minor refactorings.

I've left the `if(_buf.size() == 0)` as a follow up, because in `read_exactly_part` there is at the beginning the allocation of the tmp_buf of size `n`, which would be redundant in the `if(_buf.size() == 0)` branch in `read_exactly` in case of EOF. Does that make sense?